source/MaterialXRuntime/Private/PvtAttribute.cpp

```diff
@@ -35,31 +36,53 @@ PvtOutput::PvtOutput(const RtToken& name, const RtToken& type, uint32_t flags, P
 
 bool PvtOutput::isConnectable(const PvtInput* other) const
 {
-    // TODO: Check if the data types are compatible.
-    return _parent != other->_parent;
+    // We cannot connect to ourselves.
+    if (_parent == other->_parent)
+    {
+        return false;
+    }
+    // We must have matching types.
+    if (getType() != other->getType())
+    {
+        // TODO: Check if the data types are compatible/castable
+        // instead of strictly an exact match.
+        return false;
+    }
+    return true;
 }
 
 void PvtOutput::connect(PvtInput* input)
 {
+    // Check if the connection exists already.
+    if (input->_connection == hnd())
+    {
+        return;
+    }
+
+    // Validate the connection.
     if (input->isConnected())
     {
-        throw ExceptionRuntimeError("Input is already connected");
+        throw ExceptionRuntimeError("Input '" + input->getPath().asString() + "' is already connected");
     }
     if (!isConnectable(input))
     {
-        throw ExceptionRuntimeError("Output is not connectable to this input");
+        throw ExceptionRuntimeError("Output '" + getPath().asString() + "' is not connectable to input '" + input->getPath().asString() + "'");
     }
+
+    // Make the connection.
     _connections.push_back(input->hnd());
     input->_connection = hnd();
 }
 
 void PvtOutput::disconnect(PvtInput* input)
 {
+    // Check if the connection exists.
     if (!input->_connection || input->_connection.get() != this)
     {
-        throw ExceptionRuntimeError("Given source and destination is not connected");
+        return;
```

Comment:
Just curious, why is this not an exception? Won't users be surprised to learn that they didn't do anything?

Reply:
The operation here is to disconnect the ports. If they are not connected we have nothing to disconnect, and we can juat silently return. The user should be happy because the operation was successfully in the sense that the ports are disconnected (still).

I changed this partly for consistency with the _connect_ case above, were we do the same thing of not reporting if the connection already exists. But also because it feels a bit harsh to throw an exception if the user requests a state change and that state was already set.